### PR TITLE
Drop static RouteInfo matched method

### PR DESCRIPTION
**Pull request by wdalmut**
Just dropped out the `matched` static method in favor of default
constructor.

### CODE REVIEW

File: src/Dispatcher.php [hunk 1]
```diff
@@ -11,6 +11,7 @@
 use Penny\Exception\MethodNotAllowed;
 use Penny\Exception\RouteNotFound;
 use Psr\Http\Message\RequestInterface;
+use Penny\Route\RouteInfo;
 
 class Dispatcher
```
Comment (wdalmut) [inline src/Dispatcher.php:14]: This a `Router` right? Could `Router` be a better name?

File: src/Route/RouteInfo.php [hunk 2]
```diff
@@ -1,19 +1,17 @@
 <?php
 namespace Penny\Route;
 
-class FastPsr7RouteInfo implements RouteInfoInterface
```
Comment (wdalmut) [inline src/Route/RouteInfo.php:4]: Why this name? This is not related to **psr7** right?

File: src/Route/RouteInfo.php [hunk 3]
```diff
@@ -1,19 +1,17 @@
...
+class RouteInfo implements RouteInfoInterface
 {
     private $name;
     private $callable;
     private $params;
 
-    public static function matched($name, callable $callable, $params = [])
```
Comment (wdalmut) [inline src/Route/RouteInfo.php:10]: There is no reason to use a static factory method right?

File: src/Route/RouteInfoInterface.php [hunk 4]
```diff
@@ -3,7 +3,6 @@
 
 interface RouteInfoInterface
 {
-    public static function matched($name, callable $callable, $params);
```
Comment (wdalmut) [inline src/Route/RouteInfoInterface.php:6]: The interface doesn't need this factory method

File: tests/Utils/FastSymfonyDispatcher.php [hunk 5]
```diff
@@ -2,9 +2,10 @@
...
 namespace PennyTest\Utils;
 
-use Symfony\Component\HttpFoundation\Request;
-use Penny\Route\FastPsr7RouteInfo;
 use ReflectionClass;
+use Penny\Route\RouteInfo;
+use Penny\Route\FastPsr7RouteInfo;
+use Symfony\Component\HttpFoundation\Request;
 
 class FastSymfonyDispatcher
```
Comment (wdalmut) [inline tests/Utils/FastSymfonyDispatcher.php]: This should be a `SymfonyRouter` right?

File: src/Route/RouteInfo.php (hunk 2 above)
Comment (gianarb) [inline src/Route/RouteInfo.php:4]: Yes I'm crazy :) Can you change it in FastRouteInfo :)

File: src/Dispatcher.php (hunk 1 above)
Comment (gianarb) [inline src/Dispatcher.php:14]: mm IMO this is a Route.. not a Route_r_.. :)

File: src/Route/RouteInfo.php (hunk 2 above)
Comment (wdalmut) [inline src/Route/RouteInfo.php:4]: Why fast?

Comment (wdalmut) [inline src/Route/RouteInfo.php:4]: It is just a `RouteInfo`

File: src/Dispatcher.php [hunk 6]
```diff
@@ -6,27 +6,27 @@
 use Exception;
 use ReflectionClass;
 use FastRoute\Dispatcher as BaseDispatcher;
-use FastRoute\Dispatcher as FastRouterDispatcherInterface;
```
Comment (wdalmut) [inline src/Dispatcher.php:9]: Why this is duplicated?

File: src/Route/RouteInfo.php (hunk 2 above)
Comment (wdalmut) [inline src/Route/RouteInfo.php:4]: isn't it?

File: src/Dispatcher.php (hunk 1 above)
Comment (wdalmut) [inline src/Dispatcher.php:14]: Don't know... It just return a `RouteInfo` there is no dispatch in place, right?

File: src/Dispatcher.php (hunk 6 above)
Comment (gianarb) [inline src/Dispatcher.php:9]: Misunderstanding

File: src/Route/RouteInfoInterface.php (hunk 4 above)
Comment (gianarb) [inline src/Route/RouteInfoInterface.php:6]: We can replace this method with __constructor

File: src/Route/RouteInfo.php (hunk 3 above)
Comment (gianarb) [inline src/Route/RouteInfo.php:10]: correct

File: src/Route/RouteInfo.php (hunk 2 above)
Comment (gianarb) [inline src/Route/RouteInfo.php:4]: eheh yes agree.. This is a RouteInfo == info about current matched route

File: src/Dispatcher.php (hunk 6 above)
Comment (gianarb) [inline src/Dispatcher.php:9]: We can remove `as`.

File: src/Dispatcher.php (hunk 1 above)
Comment (gianarb) [inline src/Dispatcher.php:14]: Oh yes sorry.. I confused line.. Yes we can call it Router..

> The `Router` `dispatches` a request and return `RouterInfo`

is it a good idea? :)

File: src/Dispatcher.php [hunk 7]
```diff
@@ -65,7 +65,7 @@ public function __invoke(RequestInterface $request)
                 $function = (new ReflectionClass($controller))->getShortName();
 
                 $eventName = sprintf('%s.%s', strtolower($function), $method);
```
Comment (wdalmut) [inline src/Dispatcher.php]: This could be replaced with 

``` php
$function = strtolower((new ReflectionClass($controller))->getShortName());
$eventName = "{$function}.{$method}"; // this improve ~1us
```

| TIME | Total 1e5 | Single |
| --- | --- | --- |
| TIME | 1.500784 | 0.015008 [ms] |
| TIME | 1.411284 | 0.014113 [ms] |

(no reason) :smile: 

I have also tried with a more ugly version :smile_cat: :

``` php
$function = strtolower(basename(str_replace("\\", "/", get_class($controller))));
```

But there is no performance improvements

File: src/Dispatcher.php (hunk 1 above)
Comment (wdalmut) [inline src/Dispatcher.php:14]: don't know effectively is a wrapper around `FastRouterDispatcherInterface` interface in order... Maybe we can let this as Dispatcher for now...

Comment (wdalmut) [inline src/Dispatcher.php]: Just for records:

```php
<?php
// route-bench.php
use Penny\Dispatcher;
use Interop\Container\ContainerInterface;
use Zend\Diactoros\Request;

require __DIR__ . '/vendor/autoload.php';
require __DIR__ . '/NewDispatcher.php';

class Container implements ContainerInterface
{
    public function get($id)
    {
        return $this;
    }

    public function has($id)
    {
        return true;
    }

    public function stub() { }
}
$container = new Container();

class Router implements \FastRoute\Dispatcher
{
    public function dispatch($httpMethod, $uri)
    {
        return [\FastRoute\Dispatcher::FOUND, [$this, "stub"], []];
    }

    public function stub() { }
}
$router = new Router();

$start = microtime(true);
$dispatcher = new Dispatcher($router, $container);
$request = new Request();
for ($i=0; $i<1e5; $i++) {
    $dispatcher($request);
}
$end = microtime(true);

echo sprintf("TIME: %f, %f [ms]\n", $end-$start, ($end-$start)/1e5*1e3);


$start = microtime(true);
$dispatcher = new Penny\NewDispatcher($router, $container);
$request = new Request();
for ($i=0; $i<1e5; $i++) {
    $dispatcher($request);
}
$end = microtime(true);

echo sprintf("TIME: %f, %f [ms]\n", $end-$start, ($end-$start)/1e5*1e3);
```

and a new `NewDispatcher`

```php
<?php
namespace Penny;

use Interop\Container\ContainerInterface;
use Exception;
use ReflectionClass;
use FastRoute\Dispatcher as BaseDispatcher;
use Penny\Route\FastPsr7RouteInfo;
use Penny\Exception\MethodNotAllowed;
use Penny\Exception\RouteNotFound;
use Psr\Http\Message\RequestInterface;
use Penny\Route\RouteInfo;

class NewDispatcher
{
    /**
     * Inner dispatcher.
     *
     * @var BaseDispatcher
     */
    private $router;

    /**
     * Class constructor with required FastRoute dispatcher implementation.
     *
     * @param BaseDispatcher $router Inner router (based on Nikic FastRouter).
     */
    public function __construct(BaseDispatcher $router, ContainerInterface $container)
    {
        $this->router = $router;
        $this->container = $container;
    }

    /**
     * Dispatching.
     *
     * @param RequestInterface $request Representation of an outgoing,
     *                                  client-side request.
     *
     * @throws RouteNotFound    If the route is not found.
     * @throws MethodNotAllowed If the method is not allowed.
     * @throws Exception        If no one case is matched.
     *
     * @return array
     */
    public function __invoke(RequestInterface $request)
    {
        $router = $this->router;
        $uri = $request->getUri();

        $routeInfo = $router->dispatch(
            $request->getMethod(),
            $uri->getPath()
        );

        switch ($routeInfo[0]) {
            case BaseDispatcher::NOT_FOUND:
                throw new RouteNotFound();
            case BaseDispatcher::METHOD_NOT_ALLOWED:
                throw new MethodNotAllowed();
            case BaseDispatcher::FOUND:
                $controller = $this->container->get($routeInfo[1][0]);
                $method = $routeInfo[1][1];
                $function = strtolower((new ReflectionClass($controller))->getShortName());

                $eventName = "{$function}.{$method}";
                $routeInfo = new RouteInfo($eventName, [$controller, $routeInfo[1][1]], $routeInfo[2]);
                return $routeInfo;
            default:
                throw new Exception(null, 500);
        }
    }
}
```

File: src/Dispatcher.php (hunk 7 above)
Comment (wdalmut) [inline src/Dispatcher.php]: Just for records:

``` php
<?php
// route-bench.php
use Penny\Dispatcher;
use Interop\Container\ContainerInterface;
use Zend\Diactoros\Request;

require __DIR__ . '/vendor/autoload.php';
require __DIR__ . '/NewDispatcher.php';

class Container implements ContainerInterface
{
    public function get($id)
    {
        return $this;
    }

    public function has($id)
    {
        return true;
    }

    public function stub() { }
}
$container = new Container();

class Router implements \FastRoute\Dispatcher
{
    public function dispatch($httpMethod, $uri)
    {
        return [\FastRoute\Dispatcher::FOUND, [$this, "stub"], []];
    }

    public function stub() { }
}
$router = new Router();

$start = microtime(true);
$dispatcher = new Dispatcher($router, $container);
$request = new Request();
for ($i=0; $i<1e5; $i++) {
    $dispatcher($request);
}
$end = microtime(true);

echo sprintf("TIME: %f, %f [ms]\n", $end-$start, ($end-$start)/1e5*1e3);


$start = microtime(true);
$dispatcher = new Penny\NewDispatcher($router, $container);
$request = new Request();
for ($i=0; $i<1e5; $i++) {
    $dispatcher($request);
}
$end = microtime(true);

echo sprintf("TIME: %f, %f [ms]\n", $end-$start, ($end-$start)/1e5*1e3);
```

and a new `NewDispatcher`

``` php
<?php
namespace Penny;

use Interop\Container\ContainerInterface;
use Exception;
use ReflectionClass;
use FastRoute\Dispatcher as BaseDispatcher;
use Penny\Route\FastPsr7RouteInfo;
use Penny\Exception\MethodNotAllowed;
use Penny\Exception\RouteNotFound;
use Psr\Http\Message\RequestInterface;
use Penny\Route\RouteInfo;

class NewDispatcher
{
    /**
     * Inner dispatcher.
     *
     * @var BaseDispatcher
     */
    private $router;

    /**
     * Class constructor with required FastRoute dispatcher implementation.
     *
     * @param BaseDispatcher $router Inner router (based on Nikic FastRouter).
     */
    public function __construct(BaseDispatcher $router, ContainerInterface $container)
    {
        $this->router = $router;
        $this->container = $container;
    }

    /**
     * Dispatching.
     *
     * @param RequestInterface $request Representation of an outgoing,
     *                                  client-side request.
     *
     * @throws RouteNotFound    If the route is not found.
     * @throws MethodNotAllowed If the method is not allowed.
     * @throws Exception        If no one case is matched.
     *
     * @return array
     */
    public function __invoke(RequestInterface $request)
    {
        $router = $this->router;
        $uri = $request->getUri();

        $routeInfo = $router->dispatch(
            $request->getMethod(),
            $uri->getPath()
        );

        switch ($routeInfo[0]) {
            case BaseDispatcher::NOT_FOUND:
                throw new RouteNotFound();
            case BaseDispatcher::METHOD_NOT_ALLOWED:
                throw new MethodNotAllowed();
            case BaseDispatcher::FOUND:
                $controller = $this->container->get($routeInfo[1][0]);
                $method = $routeInfo[1][1];
                $function = strtolower((new ReflectionClass($controller))->getShortName());

                $eventName = "{$function}.{$method}";
                $routeInfo = new RouteInfo($eventName, [$controller, $routeInfo[1][1]], $routeInfo[2]);
                return $routeInfo;
            default:
                throw new Exception(null, 500);
        }
    }
}
```

File: src/Dispatcher.php (hunk 6 above)
Comment (wdalmut) [inline src/Dispatcher.php:9]: sadly the class name is already `Dispatcher` we have to let as `BaseDispatcher` or another name like: `DispatcherInterface` if you prefere

File: src/Route/RouteInfo.php (hunk 2 above)
Comment (wdalmut) [inline src/Route/RouteInfo.php:4]: ;)

File: src/Dispatcher.php (hunk 7 above)
Comment (gianarb) [inline src/Dispatcher.php]: ahah php mystery..
Replace my code with your more speedy implementation please :)

## Question for future

#96

File: src/Dispatcher.php (hunk 6 above)
Comment (gianarb) [inline src/Dispatcher.php:9]: we can maintain only 

``` php
use FastRoute\Dispatcher as BaseDispatcher;
```

:+1:

Comment (wdalmut) [inline src/Dispatcher.php:9]: :+1:

File: src/Dispatcher.php (hunk 7 above)
Comment (wdalmut) [inline src/Dispatcher.php]: Can you do something like  `$object::class` in php?

Comment (gianarb) [inline src/Dispatcher.php]: No :(
we can use `get_class($object)`

File: tests/Route/RouteInfoTest.php [hunk 8]
```diff
@@ -5,23 +5,19 @@
...
+use Penny\Route\RouteInfo;
 
 class RouteInfoTest extends PHPUnit_Framework_TestCase
 {
     public function testRouteInfoImplementInterface()
     {
-        $routeInfo = new FastPsr7RouteInfo();
-        $this->assertInstanceOf(RouteInfoInterface::class, $routeInfo);
+        $routeInfo = new RouteInfo("", function(){}, []);
+        $this->assertInstanceOf("Penny\Route\RouteInfoInterface", $routeInfo);
```
Comment (samsonasik) [inline tests/Route/RouteInfoTest.php]: single quote, and use `RouteInfoInterface::class` instead

File: src/Route/RouteInfoInterface.php (hunk 4 above)
Comment (samsonasik) [inline src/Route/RouteInfoInterface.php:6]: there should be no `__construct` in interface

File: tests/Route/RouteInfoTest.php [hunk 9]
```diff
@@ -5,23 +5,19 @@
...
 
     public function testMatched()
     {
-        $fastRouteInfo = [
-            [],
-            [new IndexController(), 'index'],
-            ["id" => 5]
-        ];
-        $routeInfo = FastPsr7RouteInfo::matched("index.try", $fastRouteInfo[1], $fastRouteInfo[2]);
+        $routeInfo = new RouteInfo("index.try", [new IndexController(), 'index'], ["id" => 5]);
```
Comment (samsonasik) [inline tests/Route/RouteInfoTest.php]: single quote also

File: tests/Utils/FastSymfonyDispatcher.php [hunk 10]
```diff
@@ -2,9 +2,10 @@
 
 namespace PennyTest\Utils;
 
-use Symfony\Component\HttpFoundation\Request;
-use Penny\Route\FastPsr7RouteInfo;
 use ReflectionClass;
+use Penny\Route\RouteInfo;
+use Penny\Route\FastPsr7RouteInfo;
```
Comment (samsonasik) [inline tests/Utils/FastSymfonyDispatcher.php]: remove this unused use `use Penny\Route\FastPsr7RouteInfo;`

File: tests/Route/RouteInfoTest.php (hunk 8 above)
Comment (wdalmut) [inline tests/Route/RouteInfoTest.php]: :+1:

**Comment by gianarb** [comment]
Thanks :+1:
